File: android/src/main/kotlin/com/example/multithread_downloads/DownloadModels.kt
package com.example.multithread_downloads

import kotlinx.coroutines.Job

data class DownloadTask(
    val url: String,
    var filePath: String,
    val fileName: String,
    val headers: Map<String, String>,
    val retryCount: Int,
    val timeoutSeconds: Int,
    var totalBytes: Long = 0,
    var downloadedBytes: Long = 0,
    var status: DownloadStatus = DownloadStatus.PENDING,
    var error: String? = null,
    var startTime: Long = 0,
    var lastSpeedUpdate: Long = 0,
    var job: Job? = null,
    var speedHistory: MutableList<Double> = mutableListOf()
)

enum class DownloadStatus(val value: Int) {
    PENDING(0), DOWNLOADING(1), PAUSED(2), COMPLETED(3), FAILED(4), CANCELLED(5)
}

data class SegmentTask(
    val url: String,
    val fileName: String,
    val size: Long = 0,
    var downloaded: Boolean = false,
    var bytes: Long = 0,
    val duration: Double = 10.0
)

data class VariantPlaylist(
    val url: String,
    val fileName: String,
    val bandwidth: Long = 0,
//    val segments: MutableList<SegmentTask> = mutableListOf(),
    val resolution: String = ""
)

File: ios/Classes/DownloadModels.swift
import Foundation

@available(iOS 13.0, *)
class MTDownloadTask {
    var url: String
    var filePath: String
    var fileName: String
    var headers: [String: String]

    var retryCount: Int = 3
    var totalBytes: Int64 = 0
    var downloadedBytes: Int64 = 0
    var status: MTDownloadStatus = .pending
    var error: String? = nil
    var startTime: Double = 0
    var lastSpeedUpdate: Double = 0
    var job: Task<Void, Error>? = nil
    var speedHistory: [Double] = []
    var timeoutSeconds: Int = 30

    // ✅ Required initializer
    init(url: String, filePath: String, fileName: String, headers: [String: String]) {
        self.url = url
        self.filePath = filePath
        self.fileName = fileName
        self.headers = headers
    }
}


enum MTDownloadStatus: Int {
    case pending = 0
    case downloading = 1
    case paused = 2
    case completed = 3
    case failed = 4
    case cancelled = 5
    case initializing = 6
}
